### PR TITLE
Endpoints for Creating stocks and trading

### PR DESCRIPTION
**Pull request by sumitd94**
# Stock Exchange

* [Here](https://github.com/Real-Dev-Squad/website-api-contracts/tree/main/crypto/stock-exchange) is the API Contracts for this
* [Here](https://whimsical.com/stock-exchange-crypto-SjagQZTj1gBHF4dtJvm818) is the link to the flow chart of the initial flow for trading.

## This PR Includes the endpoints for creating and fetching stocks data and allow the users to trade
- [x] POST request to Create new stocks
- [x] GET request to fetch all the stocks
- [x] PATCH request for trading
- [ ] Tests (Since this is high priority feature, the tests will be coming in another PR. The issue is created and I have assigned it to myself #235

### CODE REVIEW

File: routes/trading.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+const { trade } = require('../controllers/tradingController')
+const { newTrade } = require('../middlewares/validators/trading')
+
+/**
+ * @swagger
+ * /trade:
+ *  patch:
+ *   summary: Used for new trading request
+ *   tags:
+ *     - Trading
```
Comment (Rucha1499) [inline routes/trading.js:13]: We need to write this as we are using cookies, right?
` security:
       - bearerAuth: []`

**Review by Rucha1499 — COMMENTED** [review]
Great work! @sumitd94  🔥 
Requesting a minor change. Thank you! 😃

File: controllers/stocksController.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * @param req {Object} - Express request object
+ * @param req.body {Object} - Stock object
+ * @param res {Object} - Express response object
+ */
+const addNewStock = async (req, res) => {
+  try {
+    const { id, stockDetails } = await stocks.addStock(req.body)
+    return res.json({
+      message: 'Stock created successfully!',
+      stock: stockDetails,
```
Comment (swarajpure) [inline controllers/stocksController.js:14]: No validation on `req.body`?

File: controllers/stocksController.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * Fetches all the stocks
+ *
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+const fetchStocks = async (req, res) => {
+  try {
+    const allStock = await stocks.fetchStocks()
+    return res.json({
+      message: 'stocks returned successfully!',
```
Comment (swarajpure) [inline controllers/stocksController.js:32]: NIT: `Stocks` instead of `stocks`

File: models/stocks.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ * Adds Stocks
+ *
+ * @param stockData { Object }: stock data object to be stored in DB
+ * @return {Promise<{stockId: string}>}
+ */
+const addStock = async (stockData) => {
+  try {
+    const { id } = await stocksModel.add(stockData)
+    const { stockDetails } = await fetchStock(id)
+    return { id, stockDetails }
```
Comment (swarajpure) [inline models/stocks.js:13]: Why are we making 2 DB calls? Can we just not return what we stored in the DB?

File: routes/stocks.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ *           schema:
+ *             $ref: '#/components/schemas/stocks'
+ *     500:
+ *       description: badImplementation
+ *       content:
+ *         application/json:
+ *           schema:
+ *             $ref: '#/components/schemas/errors/badImplementation'
+ */
+router.post('/', createStock, addNewStock)
```
Comment (swarajpure) [inline routes/stocks.js:57]: No validation?

Comment (swarajpure) [inline routes/stocks.js:57]: No authentication?

**Review by swarajpure — CHANGES_REQUESTED** [review]
Great job in there!

Some doubts, please clarify

File: routes/stocks.js (hunk 5 above)
Comment (sumitd94) [reply routes/stocks.js:57]: `createStock` is a validation function.

File: routes/stocks.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ *         schema:
+ *           $ref: '#/components/schemas/stocks'
+ *   responses:
+ *     200:
+ *       description: returns newly created stock
+ *       content:
+ *         application/json:
+ *           schema:
+ *             $ref: '#/components/schemas/stocks'
+ *     500:
```
Comment (ankurnarkhede) [inline routes/stocks.js:50]: 401?

File: controllers/stocksController.js (hunk 2 above)
Comment (sumitd94) [reply controllers/stocksController.js:14]: We have middleware in place for validations before reaching out to controller.

File: routes/stocks.js (hunk 5 above)
Comment (ankurnarkhede) [inline routes/stocks.js:57]: Only members should be able to add a stock, right?

Comment (sumitd94) [reply routes/stocks.js:57]: We will have it once we have this [feature](https://github.com/Real-Dev-Squad/website-backend/pull/157) in place.
Only authorized app owners will be allowed to create stocks/tasks.

File: routes/trading.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
+const express = require('express')
+const router = express.Router()
+const authenticate = require('../middlewares/authenticate')
+const { trade } = require('../controllers/tradingController')
+const { newTrade } = require('../middlewares/validators/trading')
+
+/**
+ * @swagger
+ * /trade:
```
Comment (ankurnarkhede) [inline routes/trading.js:9]: Can you also modify the API schema?

File: models/stocks.js (hunk 4 above)
Comment (sumitd94) [reply models/stocks.js:13]: Ok, let me check on this.

File: routes/trading.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+ *         schema:
+ *           $ref: '#/components/schemas/trading'
+ *   responses:
+ *     200:
+ *       description: Successful trading details
+ *       content:
+ *         application/json:
+ *           schema:
+ *             $ref: '#/components/schemas/tradingSuccess'
+ *     403:
```
Comment (ankurnarkhede) [inline routes/trading.js:27]: We are returning 401 from the authenticated middleware.

**Review by ankurnarkhede — CHANGES_REQUESTED** [review]

File: routes/stocks.js (hunk 5 above)
Comment (sumitd94) [reply routes/stocks.js:57]: https://github.com/Real-Dev-Squad/website-backend/pull/222#discussion_r592711694

File: routes/trading.js (hunk 7 above)
Comment (sumitd94) [reply routes/trading.js:9]: the one in `public/apiSchema.json` ?

File: routes/trading.js (hunk 8 above)
Comment (sumitd94) [reply routes/trading.js:35]: Added this - https://github.com/Real-Dev-Squad/website-backend/pull/222/commits/9ae954a9bd04ab81abdfab5fb8a8b5b5e902aa5f
we will also be returning `403` in one case

File: routes/stocks.js (hunk 6 above)
Comment (sumitd94) [reply routes/stocks.js:56]: Done - https://github.com/Real-Dev-Squad/website-backend/pull/222/commits/9ae954a9bd04ab81abdfab5fb8a8b5b5e902aa5f

File: routes/trading.js (hunk 7 above)
Comment (sumitd94) [reply routes/trading.js:9]: https://github.com/Real-Dev-Squad/website-backend/pull/222/commits/7de13d33fef4497c26f6921780a28ce426b18a20

File: controllers/stocksController.js (hunk 3 above)
Comment (sumitd94) [reply controllers/stocksController.js]: https://github.com/Real-Dev-Squad/website-backend/pull/222/commits/7de13d33fef4497c26f6921780a28ce426b18a20

File: controllers/tradingController.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+const trade = async (req, res) => {
+  try {
+    const { username } = req.params
+    if (!username) {
+      return res.boom.forbidden('Invalid username')
+    }
+
+    const { canUserTrade, errorMessage, userBalance } = await tradeModel.trade(req.body, username)
+
+    if (!canUserTrade && canUserTrade !== undefined) {
```
Comment (ankurnarkhede) [inline controllers/tradingController.js:17]: If canUserTrade === undefined, !canUserTrade willbe true. Hence, simply adding !canUserTrade will work 🤔

File: controllers/tradingController.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+      return res.boom.forbidden(errorMessage)
+    }
+
+    return res.status(200).json({ userBalance })
```
Comment (ankurnarkhede) [inline controllers/tradingController.js:21]: .status(200) is not required as it is by default.

File: routes/trading.js [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+const router = express.Router()
+const authenticate = require('../middlewares/authenticate')
+const { trade } = require('../controllers/tradingController')
+const { newTrade } = require('../middlewares/validators/trading')
+
+/**
+ * @swagger
+ * /trade:
+ *  patch:
+ *   summary: Used for new trading request
```
Comment (ankurnarkhede) [inline routes/trading.js:11]: @sumitd94 , can uou run generate schema?

File: models/stocks.js (hunk 4 above)
Comment (sumitd94) [reply models/stocks.js]: https://github.com/Real-Dev-Squad/website-backend/pull/222/commits/c0e9f785e05885af8fd62562cbbcbd51626481bd

File: routes/stocks.js (hunk 5 above)
Comment (ankurnarkhede) [reply routes/stocks.js:63]: We should add a validation of is members for now.

**Review by ankurnarkhede — CHANGES_REQUESTED** [review]

File: routes/trading.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
+const express = require('express')
+const router = express.Router()
+const authenticate = require('../middlewares/authenticate')
+const { trade } = require('../controllers/tradingController')
+const { newTrade } = require('../middlewares/validators/trading')
+
+/**
+ * @swagger
+ * /trade:
```
Comment (Rucha1499) [inline routes/trading.js:9]: Can you please change this to `/trade/:username` ?

**Review by Rucha1499 — CHANGES_REQUESTED** [review]

File: models/trading.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+          return { canUserTrade: false, errorMessage: 'Trade was not successful due to insufficient funds' }
+        }
+        quantityToUpdate = stockData.quantity - qtyUserCanPurchase
+        userBalance = qtyUserCanPurchase * stockData.price - totalPrice
+
+        stockPriceToBeUpdated = getUpdatedPrice(stockData.price)
+        break
+      }
+      default: {
+        return { canUserTrade: false, errorMessage: 'Invalid tarde type' }
```
Comment (Rucha1499) [inline models/trading.js:50]: Minor typo in errorMessage : `Invalid trade type` . 😅

File: controllers/tradingController.js (hunk 9 above)
Comment (sumitd94) [reply controllers/tradingController.js]: https://github.com/Real-Dev-Squad/website-backend/pull/222/commits/c47c5f50b9d54d8fd2e4436aace96f4660f4bf9c

File: controllers/tradingController.js (hunk 10 above)
Comment (sumitd94) [reply controllers/tradingController.js]: https://github.com/Real-Dev-Squad/website-backend/pull/222/commits/c47c5f50b9d54d8fd2e4436aace96f4660f4bf9c

File: models/stocks.js (hunk 4 above)
Comment (sumitd94) [reply models/stocks.js]: Done.

File: routes/trading.js (hunk 12 above)
Comment (sumitd94) [reply routes/trading.js]: We have a new endpoint - https://github.com/Real-Dev-Squad/website-backend/pull/222/commits/5bae0772542bd974fcefb7f2c60b6519b88f8da8

**Review by swarajpure — APPROVED** [review]
Great job!

FTL rocks! ⚡

File: middlewares/authorization.js [hunk 14]
```diff
@@ -0,0 +1,14 @@
+const { isMember } = require('../models/members')
+module.exports = (req, res, next) => {
+  try {
+    const { username } = req.userData
+    const isUserMember = isMember(username)
```
Comment (ankushdharkar) [inline middlewares/authorization.js:5]: Stocks should not have been created by members, only from authorized users (currently just `ankush`)

File: middlewares/authorization.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+    if (!isUserMember) {
+      return res.boom.forbidden('You are not allowed to create stocks')
+    }
+    return next()
+  } catch (err) {
+    logger.error(err)
+    return res.boom.forbidden('You are not allowed to create stocks')
```
Comment (ankushdharkar) [inline middlewares/authorization.js:12]: Why is the `authorization` messaging tightly coupled with stocks?

**Pull request merged**